buildlog/buildlog.go

```diff
@@ -0,0 +1,103 @@
+package buildlog
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"time"
+)
+
+var readInterval time.Duration = 10 * time.Millisecond
+
+// Logger outputs logs
+type Logger interface {
+	Run()
+	Stop()
+}
+
+type log struct {
+	ctx    context.Context
+	file   io.Reader
+	writer io.Writer
+	cancel context.CancelFunc
+}
+
+type logLine struct {
+	Time     int64  `json:"t"`
+	Message  string `json:"m"`
+	Line     int    `json:"n"`
+	StepName string `json:"s"`
+}
+
+// New creates new Logger interface.
+func New(ctx context.Context, filepath string, writer io.Writer) (Logger, error) {
+	log := log{
+		writer: writer,
+	}
+
+	var err error
+	log.file, err = os.OpenFile(filepath, os.O_RDONLY|os.O_CREATE|os.O_TRUNC, 0666)
+	if err != nil {
+		return log, fmt.Errorf("failed to open raw build log file: %w", err)
+	}
+
+	log.ctx, log.cancel = context.WithCancel(ctx)
+
+	return log, nil
+}
+
+func (l log) Stop() {
+	l.cancel()
+	return
+}
+
+func (l log) Run() {
+	reader := bufio.NewReader(l.file)
+
+	for {
+		select {
+		case <-l.ctx.Done():
+			break
+		default:
+			err := l.output(reader)
+			if err != nil {
+				fmt.Printf("failed to run logger: %w\n", err)
+				l.cancel()
```

Comment:
Can we cancel with every error?
I think it stop logging, when failed to parse log.

Reply:
Or can we add the function of continue writing log at second pass?

Reply:
> Can we cancel with every error?
> I think it stop logging, when failed to parse log.

I think it work that way.

> Or can we add the function of continue writing log at second pass?

If broken line is come, logger is stopped, but build is running.
So we have no plan to do so for now.

Reply:
I see. Someday I want these function. Users are hard to understand what happen, if there are no formatted log even the build is running.